py/server/deephaven/experimental/partitioned_table_service.py

```diff
@@ -0,0 +1,326 @@
+                       callback: Callable[[int], None]) -> None:
+        """ Provides a callback for the backend service to pass the size of the partition with the given table key
+        and partition location key. The callback should be called with the size of the partition in number of rows.
+
+        This is called for tables created when ：meth:`PythonTableDataService.make_table` is called with live=False.
+
+        Args:
+            table_key (TableKey): the table key
+            table_location_key (PartitionedTableLocationKey): the partition location key
+            callback (Callable[[int], None]): the callback function
+        """
+        pass
+
+    @abstractmethod
+    def subscribe_to_partition_size_changes(self, table_key: TableKey, table_location_key: PartitionedTableLocationKey,
+                                            callback: Callable[[int], None]) -> Callable[[], None]:
+        """ Provides a callback for the backend service to pass the changed size of the partition with the given
+        table key and partition location key. The callback should be called with the size of the partition in number of
+        rows.
+
+        This is called for tables created when ：meth:`PythonTableDataService.make_table` is called with live=True
+
+        Note that the callback must be called with the initial size of the partition after this method has returned and
+        must not be called before this method has returned.
+
+        The return value is a function that can be called to unsubscribe from the partition size changes.
+
+        Args:
+            table_key (TableKey): the table key
+            table_location_key (PartitionedTableLocationKey): the partition location key
+            callback (Callable[[int], None]): the callback function
+
+        Returns:
+            Callable[[], None]: a function that can be called to unsubscribe from the partition size changes
+        """
+        pass
+
+    @abstractmethod
+    def column_values(self, table_key: TableKey, table_location_key: PartitionedTableLocationKey, col: str, offset: int,
+                      min_rows: int, max_rows: int) -> pa.Table:
+        """ Returns the values for the column with the given name for the partition with the given table key and
+        partition location key. The returned pa.Table should have a single column with values of the specified range
+        requirement for the given column.
+
+        Args:
+            table_key (TableKey): the table key
+            table_location_key (PartitionedTableLocationKey): the partition location key
+            col (str): the column name
+            offset (int): the starting row index
+            min_rows (int): the minimum number of rows to return
+            max_rows (int): the maximum number of rows to return
+
+        Returns:
+            pa.Table: a pa.Table that contains the values for the column
+        """
+        pass
+
+
+class PythonTableDataService(JObjectWrapper):
```

Comment:
I'm confused about what is going on.  This method has a single public interface method, and it is to create a new table.  No data is added to the table, etc.  It doesn't sound useful.  I'm clearly missing something.

Reply:
A `TableDataService` is the Java API for providing the guts of `SourceTable` instances in Java. It provides `TableLocationProvider` instances, which in turn provide the `TableLocation` instances that map to the regions of a `Table` if they aren't filtered out when the table is `coalesce()`ed.

This implementation is a bridge allowing a Python shop that has data in Arrow Flight to offer a `TableDataService`, and the `makeTable()` method is just doing the glue for users to get a `Table` out.

This is a systems programmer's interface, not an end-user's interface.